src/DataStructures/LeviCivitaIterator.cpp

```diff
@@ -0,0 +1,68 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "DataStructures/LeviCivitaIterator.hpp"
+
+#include <algorithm>
+#include <numeric>
+
+#include "DataStructures/Index.hpp"
+#include "Utilities/Gsl.hpp"
+
+namespace {
+template <size_t Dim>
+constexpr void initialize_indexes_and_signs(
+    gsl::not_null<std::array<Index<Dim>, factorial(Dim)>*> indexes,
+    gsl::not_null<std::array<int, factorial(Dim)>*> signs) noexcept {
+  auto index = Index<Dim>();
+  std::iota(index.begin(), index.end(), 0);
+
+  // Add the initial index permutation to the list
+  // Also add the sign, which is always +1, because std::iota
+  // initializes the indices in index as 0,1,...Dim-1
+  (*indexes)[0] = index;
+  (*signs)[0] = 1;
+
+  size_t permutation_count = 1;
+  auto factoradic_counter = std::array<size_t, Dim>();
+
+  // std::next_permutation generates the different permuations of index,
+  // but how do you know whether the corresponding Levi Civita symbol is
+  // +1 or -1? To find out, compute the number, in the factoradic number system,
+  // corresponding to each permutation, and sum the digits. If the sum is
+  // even (odd), the corresponding Levi-Civita symbol will be +1 (-1).
+  // This works because there are Dim! unique permutations of the
+  // Levi Civita symbol indices, so each one can be represented by a
+  // (`Dim`)-digit factorial-number-system number. For more on the
+  // factoradic number system, see
+  // https://en.wikipedia.org/wiki/Factorial_number_system
+  for (; std::next_permutation(index.begin(), index.end());) {
+    for (size_t i = 0; i < Dim; ++i) {
+      gsl::at(factoradic_counter, i)++;
+      if (gsl::at(factoradic_counter, i) < i + 1) {
+        break;
+      } else {
+        gsl::at(factoradic_counter, i) = 0;
+      }
+    }
+
+    gsl::at((*indexes), permutation_count) = index;
+    gsl::at((*signs), permutation_count) =
+        (std::accumulate(factoradic_counter.begin(), factoradic_counter.end(),
+                         size_t(0)) %
+         2) == 0
+            ? 1
+            : -1;
+    ++permutation_count;
+  }
+}
+}  // namespace
+
+template <size_t Dim>
+LeviCivitaIterator<Dim>::LeviCivitaIterator() noexcept {
+  initialize_indexes_and_signs<Dim>(&indexes_, &signs_);
+}
+
+template class LeviCivitaIterator<2>;
```

Comment:
Instantiate for 1 and if possible 0 to allow use in generic code.

Reply:
Sure, if you like (for 1). I’ve never seen LeviCivita defined for dim < 2. For 1, there is only one term, eps0=1, so all you are doing is multiplying by 1. Why would you ever really want do this, as opposed to a new developer doing it by accident? For 0, eps is undefined, so there is no way to instantiate that case.

Reply:
So I can write something like
```c++
template <size_t Dim>
double volume_element(const std::array<Tensor<...>, Dim>& basis) {
  double result = 0.;
  for (LeviCivitaIterator<Dim> it; it; ++it) {
    result += ...;
  }
  return result;
}
```
and have it work in all dimensions.

In 0d it is a scalar equal to 1, but we probably won't need that one so I'm fine leaving it out for now.